# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: testing/web-platform/tests/editing/other/formatblock-preserving-selection.tentative.html
# Commit: d554ca7850eb
# Full Hash: d554ca7850eb362a0482c9f4c19caf288e65efb1
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-08-04 04:34:13
# Regressor Bug: 1774704
# File Overlap Count: 1
# Description:
#   Bug 1774704 - part 0: Add automated tests of preserving selection while handling some commands r=m_kato
#   
#   The following patches touch the logic to restore selection after handling
#   commands.  However, it seems that they are not tested because I found some
#   regressions by manual testing, but didn't cause orange on tryserver.
# ==============================================================================

diff -r 86da8f2a28ea -r d554ca7850eb testing/web-platform/tests/editing/other/formatblock-preserving-selection.tentative.html
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/testing/web-platform/tests/editing/other/formatblock-preserving-selection.tentative.html	Thu Aug 04 00:21:53 2022 +0000
@@ -0,0 +1,136 @@
+<!doctype html>
+<html>
+<head>
+<meta chareset="utf-8">
+<meta name="timeout" content="long">
+<meta name="variant" content="?styleWithCSS=false&block=address">
+<meta name="variant" content="?styleWithCSS=false&block=article">
+<meta name="variant" content="?styleWithCSS=false&block=blockquote">
+<meta name="variant" content="?styleWithCSS=false&block=dd">
+<meta name="variant" content="?styleWithCSS=false&block=div">
+<meta name="variant" content="?styleWithCSS=false&block=dt">
+<meta name="variant" content="?styleWithCSS=false&block=h1">
+<meta name="variant" content="?styleWithCSS=false&block=li">
+<meta name="variant" content="?styleWithCSS=false&block=pre">
+<meta name="variant" content="?styleWithCSS=true&block=address">
+<meta name="variant" content="?styleWithCSS=true&block=article">
+<meta name="variant" content="?styleWithCSS=true&block=blockquote">
+<meta name="variant" content="?styleWithCSS=true&block=dd">
+<meta name="variant" content="?styleWithCSS=true&block=div">
+<meta name="variant" content="?styleWithCSS=true&block=dt">
+<meta name="variant" content="?styleWithCSS=true&block=h1">
+<meta name="variant" content="?styleWithCSS=true&block=li">
+<meta name="variant" content="?styleWithCSS=true&block=pre">
+<title>Test preserving selection after formatBlock</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+<script src="/resources/testdriver-actions.js"></script>
+<script src="../include/editor-test-utils.js"></script>
+</head>
+<body>
+<div contenteditable></div>
+<script>
+"use strict";
+
+const editor = document.querySelector("div[contenteditable]");
+const utils = new EditorTestUtils(editor);
+const searchParams = new URLSearchParams(document.location.search);
+const styleWithCSS = searchParams.get("styleWithCSS");
+const block = searchParams.get("block");
+document.execCommand("styleWithCSS", false, styleWithCSS);
+
+// Note that it's not scope of this test how browsers to convert the selected
+// content to a block.
+
+// html: Initial HTML which will be set editor.innerHTML, it should contain
+//       selection range with a pair of "[" or "{" and "]" or "}".
+// expectedSelectedString: After executing "outdent", compared with
+//                         getSelection().toString().replace(/[ \n\r\t]+/g, "")
+const tests = [
+  {
+    html: "a[b]c",
+    expectedSelectedString: "b",
+  },
+  {
+    html: "a[bc<br>de]f",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<div>a[b]c</div>",
+    expectedSelectedString: "b",
+  },
+  {
+    html: "<div>a[bc</div><div>de]f</div>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<div>a[bc<br>de]f</div>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<ul><li>a[b]c</li></ul>",
+    expectedSelectedString: "b",
+  },
+  {
+    html: "<ul><li>a[bc</li><li>de]f</li></ul>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<ul><li>a[bc</li><li>de]f</li><li>ghi</li></ul>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<ul><li>abc</li><li>d[ef</li><li>gh]i</li></ul>",
+    expectedSelectedString: "efgh",
+  },
+  {
+    html: "<ul><li>abc</li><li>d[ef</li></ul>" +
+          "<div>gh]i</div>",
+    expectedSelectedString: "efgh",
+  },
+  {
+    html: "<div>a[bc</div>" +
+          "<ul><li>de]f</li><li>ghi</li></ul>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<table><tr><td>a[b]c</td></tr></table>",
+    expectedSelectedString: "b",
+  },
+  {
+    html: "<table><tr><td>a[bc</td><td>de]f</td></tr></table>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<table><tr><td>a[bc</td></tr><tr><td>de]f</td></tr></table>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<div>a[bc</div>" +
+          "<table><tr><td>de]f</td></tr></table>",
+    expectedSelectedString: "bcde",
+  },
+  {
+    html: "<table><tr><td>a[bc</td></tr></table>" +
+          "<div>de]f</div>",
+    expectedSelectedString: "bcde",
+  },
+];
+
+for (const t of tests) {
+  test(() => {
+    utils.setupEditingHost(t.html);
+    document.execCommand("formatblock", false, block);
+    assert_equals(
+      getSelection().toString().replace(/[ \n\r\t]+/g, ""),
+      t.expectedSelectedString,
+      `Result: ${editor.innerHTML}`
+    );
+  }, `Preserve selection after formatBlock with ${block} at ${t.html}`);
+}
+
+</script>
+</body>
+</html>